proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+- `MigrationStatus.IGNORE` makes compiler to ignore the annotation completely
+
+
+Inappropriate usages in Kotlin that are subjects to report errors/warnings
+include:
+- Passing nullable arguments for parameters those types are enhanced from
+annotations to not null
+- Using enhanced nullable values as an argument for kotlin not-nullable 
+parameter or as a receiver for a method call
+- Overriding enhanced Java methods with incorrect signature in Kotlin
+
+```java
+@TypeQualifierNickname
+@Nonnull(when = When.ALWAYS)
+@Migration(status = MigrationStatus.WARNING)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNonnull {
+}
+
+
+interface A {
+    void foo(@MyNonnull String x);
+}
+
+```
+
+```kotlin
+fun bar(a: A) {
+    a.foo(null) // a warning will be reported on 'null' argument
+}
+```
+
+In the example the warning will be reported until status argument is changed
+to `ERROR`, then it should become an error.
+
+It's important that migration annotation works only for its immediate usages
+on types or through `TypeQualifierDefault` annotation.
+It means that another nickname annotation aliased to the one under migration
+doesn't inherit its migration status, thus by default, it would be a
+`MigrationStatus.ERROR`
+
+### Migration-related compiler flags
+
+Compiler flags might be useful mostly for library users who for some reasons 
+need a migration state different from the one offered by a library maintainer.
+
+They may be set in a build systems configuration files or in the IDE.  
+
+#### Global state of JSR-305 support
+The flag `-Xjsr305-annotations` has three options: ignore, enable and warn.
+
+It effectively defines the migration status behavior only for nullability 
+qualifier nicknames that haven't yet been annotated with `kotlin.Migration`.
+
+This annotation is necessary, since using all of the custom nullability
+qualifiers, and especially `TypeQualifierDefault` is already spread among many
+well-known libraries and users may need to migrate smoothly when updating to
+the Kotlin compiler version containing JSR-305 support.
+
+### Global migration status
+The flag `-Xjsr305-annotation-migration` has the same set of options and has 
+the similar meaning, but overrides the behavior defined by the `status` argument
+for all of the `@Migration` annotated qualifiers.
+
+It might be needed in case when library users have different view on a 
+migration status for the library: both they may want to have errors while the
+official migration status is `WARNING`, and vice versa, they may wish to 
+postpone errors reporting for some time because they're not completed their 
+migration yet.
+
+### Overriding migration status for specific annotation
+Sometimes it might be necessary to manage a migration phase for a particular
+library. That could be done with the flag `-Xjsr305-user-annotation`. It
+accepts fully-qualified name of a nullability qualifier annotation and its 
+desired status in a format `<fq.name>:ignore/warn/error`.
```

Comment:
Please provide a concrete usage example of this compiler argument here in the proposal